stdenv: make symlinks that refer to the same output relative

andir opened the pull request:
###### Motivation for this change

While looking at the graph of all the outputs in my personal binary
cache it became obvious that we have a lot of self references within the
package set. That isn't an isssue by itself. However it increases the
size of the binary cache for every (reproducible) build of a package
that carries references to itself. You can no longer deduplicate the
outputs since they are all unique. One of the ways to get rid of (a few)
references is to rewrite all the symlinks that are currently used to be
relative symlinks. Two buildd of something that didn't really change but
carry each a self-reference can the be stored as the same NAR file again.

I quickly hacked together this change to see if that would yield and
success. My bash scripting skills are probably not great but so far it
seem to somewhat work.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change (only all of those in the tested set)
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @Ericson2314

andir committed:
stdenv: make symlinks that refer to the same output relative

While looking at the graph of all the outputs in my personal binary
cache it became obvious that we have a lot of self references within the
package set. That isn't an isuse by itself. However it increases the
size of the binary cache for every (reproducible) build of a package
that carries references to itself. You can no longer deduplicate the
outputs since they are all unique. One of the ways to get rid of (a few)
references is to rewrite all the symlinks that are currently used to be
relative symlinks. Two build of something that didn't really change but
carries a self-reference can the be store as the same NAR file again.

I quickly hacked together this change to see if that would yield and
success. My bash scripting skills are probably not great but so far it
seem to somewhat work.

andir commented:
Are there valid use cases where this is not desirable? I was thinking of providing an escape hatch if there are any...

7c6f434c commented:
>Are there valid use cases where this is not desirable? I was thinking of providing an escape hatch if there are any...

I guess a weird enough upstream build system of a reverse depedency could copy something to its own installation directory, which was OK when added (back when it was a plain file), works with absolute symlinks but might break for relative ones with some flags?

This sounds weird, but maybe something close is posible.

B4dM4n commented on pkgs/build-support/setup-hooks/make-symlinks-relative.sh:
```diff
@@ -0,0 +1,23 @@
+        echo "rewriting symlink $f to be relative to $prefix"
+
+        rm "$f"
+        ln -rs "$prefix/$relativeTarget" "$f"
```
The `ln` destination should be `"$prefix/$f"`.

To simplify the process this block can be replaced with these `ln` flags:
```bash
        echo "rewriting symlink $f to be relative to $prefix"
        ln -snrf "$symlinkTarget" "$prefix/$f"
```
as `ln -r` produces the minimal possible relative link.

B4dM4n reviewed (commented):
The concept sounds nice, but I found two issues in the bash script.

> Are there valid use cases where this is not desirable? I was thinking of providing an escape hatch if there are any...

I can definitely see issues with scripts that expect an absolute path from `readlink` and will be broken by this change, so being able to disable this hook would be nice.

andir commented on pkgs/build-support/setup-hooks/make-symlinks-relative.sh:
```diff
@@ -0,0 +1,23 @@
+    local relativeTarget
+    while IFS= read -r -d $'\0' f; do
+        symlinkTarget=$(readlink "$f")
+        if [ "${symlinkTarget:0:${#prefix}}" != "$prefix" ]; then
```
Thanks! While that technically should never happen within `/nix/store` that might be a good safeguard.

I'll add that to the next version.

conferno commented:
If we are going to find all symlinks and resolve them, it would be nice to fail the build with a meaningful error message if the target (file or dir at the location a symlink points to) does not exist.
There were a couple of bugs in nixpkgs history when dangling symlinks appeared after a blind version bump. It did not fail the build postponing the problem.

conferno commented:
> Are there valid use cases where this is not desirable? I was thinking of providing an escape hatch if there are any...
> I guess a weird enough upstream build system of a reverse depedency could copy something to its own installation directory, which was OK when added (back when it was a plain file), works with absolute symlinks but might break for relative ones with some flags? This sounds weird, but maybe something close is posible.

I have seen these bugs in the wild:
1. relative `readlink`'ed path resolved using `cwd` as the base directory
2. resolving chain of symlinks using the same base directory of the first symlink in the chain

Using absolute paths in symlinks brings some redundancy against those bugs

7c6f434c commented:
>If we are going to find all symlinks and resolve them, it would be nice to fail the build with a meaningful error message if the target (file or dir at the location a symlink points to) does not exist.

… and also to provide and ways way to opt out of this behaviour. I do have some uses for symlinking to nonexistent files; for example when something wants to write to some path that is in store (but can be redirected to a proper location via a symlink)

andir commented:
> If we are going to find all symlinks and resolve them, it would be nice to fail the build with a meaningful error message if the target (file or dir at the location a symlink points to) does not exist.
> … and also to provide and ways way to opt out of this behaviour. I do have some uses for symlinking to nonexistent files; for example when something wants to write to some path that is in store (but can be redirected to a proper location via a symlink)

Please note that I am only checking for references into the same output. So in general those shouldn't be the problem. You can still create arbitrary symlinks to directories that do not exist as long as they do not point into the output. Would that work for you?

I feel like the enforcement of valid symlinks is yet another topic compared to what I wanted to accomplish here.

7c6f434c commented:
>Please note that I am only checking for references into the same output. So in general those shouldn't be the problem. You can still create arbitrary symlinks to directories that do not exist as long as they do not point into the output. Would that work for you?

Ah, right, same-output symlinks stay broken or valid after installation, and the situation where existence of a broken symlink to a nonexistent location is preferrable to plain absence of a directory entry sounds like a small component part of an unspeakable horror.

conferno commented:
> If we are going to find all symlinks and resolve them, it would be nice to fail the build with a meaningful error message if the target (file or dir at the location a symlink points to) does not exist.
> … and also to provide and ways way to opt out of this behaviour. I do have some uses for symlinking to nonexistent files; for example when something wants to write to some path that is in store (but can be redirected to a proper location via a symlink)

Ok, then just print a warning visible in ofborg logs.

conferno commented:
> I feel like the enforcement of valid symlinks is yet another topic compared to what I wanted to accomplish here.

Yes, it is another topic, but adding the warning is +1 line (and zero performance penalty) in your hook, while making a dedicated hook for the enforcement would be a controvercial issue mainly for its performace implications

conferno commented:
> Ah, right, same-output symlinks stay broken or valid after installation, and the situation where existence of a broken symlink to a nonexistent location is preferrable to plain absence of a directory entry sounds like a small component part of an unspeakable horror.

Dangling symlinks within same output could be OK in derivations "built" by `fetchzip` (llvm sources has some among the tests), so yes, failing hard is not acceptable, but printing a warning would be nice

andir commented:
I added some of the review comments. I removed the hard failure on dangling symlinks and just left the warning in there. Does that make sense?

I think failing on invalid symlinks is really a more complicated topic that we shouldn't try to solve right now. Warning is probably okay but failing is yet another different topic from this approach. We can add it to this hook in another PR once there is consensus about the desired behavior.

andir commented:
@GrahamcOfBorg build bash

conferno commented on pkgs/build-support/setup-hooks/make-symlinks-relative.sh:
```diff
@@ -0,0 +1,28 @@
+            continue
+        fi
+
+        if [ ! -e "$symlinkTarget" ]; then
```
I think this block (lines 20-22) should preceed the block above (lines 15-18) so sanity of external symlinks will be checked too.

andir replied:
I don't think that is a good idea. A brief look at nixpkgs showed that a bunch of packages point symlinks at `/etc/$foo/` or to `/run/opengl-drivers` and those clearly do not exist in the build sandbox. Expecting every single of those packages to opt out of this is not really a great solution. This should be an incremental change and not a roadblock.

conferno replied:
it is only printing a warning.
symlinks to `/etc/$foo/` also deserve to be warned about, it is a potential problem with Nix-on-{Debian,Fedora,Gentoo,...} where `/etc/$foo/` might differ from that on NixOS

Ericson2314 commented on pkgs/build-support/setup-hooks/make-symlinks-relative.sh:
```diff
@@ -0,0 +1,28 @@
+_makeSymlinksRelative() {
+    local symlinkTarget
+
+    if [ -n "${dontRewriteSymlinks}" ]; then
```
`${dontRewriteSymlinks-}` in case it is unset

andir replied:
Fixed.

FRidh commented:
Can this go in?

andir commented:
@FRidh Looks good to me :-) If @Ericson2314 is now happy I see no reason why we can't merge this.

andir commented:
Thanks @FRidh !

Ericson2314 commented:
Woo! Sorry I didn't have the time to vet this further myself. Thanks @FRidh and @andir

Artturin commented on pkgs/build-support/setup-hooks/make-symlinks-relative.sh:
```diff
@@ -0,0 +1,28 @@
+fixupOutputHooks+=(_makeSymlinksRelative)
```
wlroots has 
```
wlroots> +++ find /nix/store/3a0xwszw8n5dzzhsgfnilvsqi4hk565s-wlroots-0.15.1-examples -type l -print0
wlroots> find: '/nix/store/3a0xwszw8n5dzzhsgfnilvsqi4hk565s-wlroots-0.15.1-examples': No such file or directory
```

because the examples output is created in postFixup while this hook runs in fixupPhase

if this is changed to `postFixupHooks` then we shouldn't get a error like that in many builds

Artturin replied:
#204692
